### ISSUE TITLE
Nested many to many relay nodes return duplicate values (non deterministic)

### ISSUE THREAD

**Issue by rsomani95** [MrThearMan/graphene-django-query-optimizer#80]
### I have read the docs thoroughly before making this bug report.

- [X] Yes

### I have read through other open issues, and my issue is not a duplicate.

- [X] Yes

### What version of the library you are using?

0.3.0

### Which python version are you using?

3.12

### What operating system are you on?

Mac

### Description

There is an issue with nested relay nodes for many to many relations. 

```graphql
{
  pagedHousingCompanies {
    edges {
      node {
        id
        developers {
          totalCount
          edges {
            node {
              id
              name
            }
          }
        }
      }
    }
  }
}
```

There's a lot of duplicates returned for `developers` - as is evident by the returned data, and the mismatch with `totalCount`:

```json
{
  "data": {
    "pagedHousingCompanies": {
      "edges": [
        {
          "node": {
            "id": "SG91c2luZ0NvbXBhbnlOb2RlOjU=",
            "developers": {
              "totalCount": 3,
              "edges": [
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo1",
                    "name": "Cole Lowery"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo1",
                    "name": "Cole Lowery"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo1",
                    "name": "Cole Lowery"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo1",
                    "name": "Cole Lowery"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo5",
                    "name": "Thomas Barrett"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo5",
                    "name": "Thomas Barrett"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo5",
                    "name": "Thomas Barrett"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo5",
                    "name": "Thomas Barrett"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo5",
                    "name": "Thomas Barrett"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo5",
                    "name": "Thomas Barrett"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo5",
                    "name": "Thomas Barrett"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo5",
                    "name": "Thomas Barrett"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo3",
                    "name": "Thomas Lawrence"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo3",
                    "name": "Thomas Lawrence"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo3",
                    "name": "Thomas Lawrence"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo3",
                    "name": "Thomas Lawrence"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo3",
                    "name": "Thomas Lawrence"
                  }
                },
                {
                  "node": {
                    "id": "RGV2ZWxvcGVyTm9kZTo3",
                    "name": "Thomas Lawrence"
                  }
                }
              ]
            }
          }
        },
        ....
```

**Comment by MrThearMan** [comment]
This error has a very simple source:

```python
count_1 = Developer.objects.count()

assert count_1  == 10  # Passes

count_2 = Developer.objects.annotate(
    foo=(
        models.Window(
            expression=RowNumber(),
            partition_by=models.F("housingcompany"),
        )
    )
).count()

assert count_2  == 10  # Fails
```

Seems like the window function is adding some duplication, I'll see what's the solution here, if not just a `qs.distinct()`.

**Comment by MrThearMan** [comment]
Ok, here's my progress so far.

The fact that the window function creates duplicates is actually what we need to do. Take the example of Housing Companies and Developers: to limit developers for each housing company to their first one, we need to annotate each housing company's develpers with the row number. See the table:

| Housing Company ID | Developer ID | Row Num |
|--------------------|--------------|---------|
| 1                  | 1            | 1       |
| 1                  | 3            | 2       |
| 1                  | 5            | 3       |
| 2                  | 5            | 1       |
| 3                  | 5            | 1       |
| 4                  | 5            | 1       |
| 5                  | 4            | 1       |
| 5                  | 7            | 2       |
| 5                  | 9            | 3       |

We have duplicated housing companies, but now we know which developers we need to select if we want to limit each housing companies developers with, for example, `first:1`.

There are a few issues with the current implementation. 

First, duplication is not ideal, since we'll be loading more data from the database than is needed. `prefetch_related` fetches only the developers that are needed, and then joins them to the housing companies they relate to, so we don't get duplicates. Ideally we would do something similar, but in practice this might be difficult, since we would need to annotate each develper row with some map that tells us the row number for each housing company it belongs to (in the current queryset).  However, we don't know the housing companies in question before the prefetching is in progress, and we can't really hook into the prefetch process, nor would I want to perform such magic, so this seems like a deadend.

Secondly, the actual SQL generate by the current window function is wrong:

```sql
SELECT
    h_id,
    d_id,
    row_number
FROM (
    SELECT * FROM (
        SELECT
            h.id AS h_id,
            hd.developer_id AS d_id,
            ROW_NUMBER() OVER (PARTITION BY hd.developer_id) AS row_number
        FROM example_housingcompany h
        LEFT OUTER JOIN example_housingcompany_developers hd ON (h.id = hd.housingcompany_id)
    )
    WHERE (
        row_number >= 0
        AND row_number <= 100
    )
)
ORDER BY h_id;
```

The issue is the `PARTITION BY` clause, it should be by `h.id` and not `hd.developer_id` to generate the table above. Since the SQL is generated by django, it'll be difficult to change this, since we would still need to perform the join to the `example_housingcompany_developers` table (so that the duplication occurs), even if we don't use it in the partitioning clause.

Thirdly, if we were to go with the window function and the duplication, it will be much less efficient than fetching just the required developers and joining in python for smaller datasets. For larger datasets, limiting items to the first 100 rows, even if the duplication bumps that up to let's say 300 rows, will be much more efficient that fetching 1000 rows. This of course depends on the dataset in question, but I'm not sure which is the best path forward there.

Now, since this is an actual bug, I do want some resolution for it as quickly as possible. One option for this is to disable the windowing altogether for now, if I don't find another solution. Your thoughts would also be appreciated.

**Comment by rsomani95** [comment]
Thanks for the detailed breakdown.

One could argue that smaller datasets (again, very use case dependent) may be okay without the optimizer altogether, whereas larger datasets might be too slow to work with without it. In my personal use case, the many to many relations for my parent objects are typically 1-40 at the most per object (how that will change in the future remains to be seen...), so I wouldn't be opposed to having windowing disabled altogether in the near term.

> This of course depends on the dataset in question, but I'm not sure which is the best path forward there.
I think there's no perfect solution that doesn't rely on "magic" that can work for all use cases. That being said, in the medium term, it might be worth constructing larger test datasets to better weigh the tradeoffs.

**Comment by MrThearMan** [comment]
Ok, it seems like I've resolved the issue, although I did end up [hacking the prefetch related mechanism](https://github.com/MrThearMan/graphene-django-query-optimizer/blob/0528fa64a30228f73ef99be4e6530c7050c91fd6/query_optimizer/prefetch_hack.py#L32), although just a little bit 😉

First, my initial analysis was a wrong. The prefetch query _actually_ looks like this (simplified for brevity again, since the code generated one id a little messy):

```sql
SELECT
    housing_company_id,
    developer_id,
    idx
FROM (
    SELECT *
    FROM (
        SELECT
            ehd.housingcompany_id AS housing_company_id,
            ed.id AS developer_id,
            (ROW_NUMBER() OVER (PARTITION BY ehd.housingcompany_id ORDER BY ed.name) - 1) AS idx,
        FROM example_developer ed
        -- Two joins?
        LEFT OUTER JOIN example_housingcompany_developers ehd ON ed.id = ehd.developer_id
        INNER JOIN example_housingcompany_developers eee ON ed.id = eee.developer_id
        WHERE eee.housingcompany_id IN (<parent_ids>)
    )
   WHERE (idx >= 0 AND idx < 100)
)
```

The issue here is the double join perfomed on the `housing_company`-`developer` through table. This happens because the prefetch adds its own INNER join in order to filter the queryset based on the parent models fetched, while our window function adds an OUTER join for the paritioning.

The `PARTITION BY` clause is fine (in fact, a similar one can be found in [django's code](https://github.com/django/django/blob/7646b9023da7e1f6f3a871959db027b3ea36eebb/django/db/models/fields/related_descriptors.py#L100-L114), but we can't rely on if unfortunately), I was just thinking about the paritioning from the wrong perspective in my previous comment. So the real issue was how we can ger rid of the extra join.

The solution for this was to use some undocumented APIs of the `django.db.models.sql.query.Query` class and patch the prefetch filtering function in order for the queryset to not add the additional join. This seems to work from my testing, but it does seem a little brittle.

I didn't add any setting to disable this, since the real use for nested connection fields is for big datasets, and as you said, those can't really live without any optimization. Smaller datasets can just use the `DjangoListField` if they don't need pagination, since it also supports filtering. Hopefully this solution is solid enough, [released on 0.4.0](https://github.com/MrThearMan/graphene-django-query-optimizer/releases/tag/v0.4.0).
